### ISSUE TITLE
Мысли, что можно улучшить в оформлении

### ISSUE THREAD

**Issue by audetv** [terratensor/kob-library-app#20]
> хочется слева: назад 3,5,10, а справа, где сейчас параграфы рядом, чтобы было вперёд 3,5,10

Есть много мыслей, что можно улучшить в оформлении. В целом на правильном пути. Есть подозрение. что для демонстрации книг нам просто нужно подобрать другой шаблон, не такой как у комментариев. Сейчас быстро пробежался по шаблонам бутстрапа, что-то ничего не попалось для лонг рида. В принципе не проблема прокачать и текущий шаблон, сделать его более книго-читаемым. 

Потестировал ещё результаты. Можно сделать одну единственную кнопку «контекст», которая будет вмещать установленное число параграфов в обе стороны, на первый взгляд достаточно по три штуки вперёд и назад. Этой кнопкой может быть прямо номер комментария. В странице выдачи контекста можно убрать заголовки, нумерацию и вообще всё остальное оформление блоков, текст разделять лишь пустой строкой, как в книге. Восприниматься будет чище и легче. 

Понятно, что сейчас страница с контекстом это просто поисковая выдача по списку номеров параграфов, но мы можем сделать для показа контекста отдельный шаблон, без разделения текста на блоки? 

Что в принципе более эффективно (или вообще возможно) — найти и установить другой шаблон для общей книжной выдачи, или доработать текущий?

_Originally posted by @iprst in https://github.com/terratensor/book-parser/issues/2#issuecomment-1612144313_

**Comment by iprst** [comment]
Самое главное, что после проделанной вами работы осталось, это разобраться со скобками, оформляющими сноски. Сейчас это самый нечитабельный элемент получается. Я так понимаю основной проблемой является то, что сноски изначально являются блоком с параграфами, и таким образом текст сноски, взятый в скобки, отбивается тегом в начале и в конце, то есть после открывающей скобки и перед закрывающей, таким образом скобки отдаляются от текста, который они окружают и по сути не связываются с ним визуально.

**Comment by audetv** [comment]
> Самое главное, что после проделанной вами работы осталось, это разобраться со скобками, оформляющими сноски. Сейчас это самый нечитабельный элемент получается. Я так понимаю основной проблемой является то, что сноски изначально являются блоком с параграфами, и таким образом текст сноски, взятый в скобки, отбивается тегом в начале и в конце, то есть после открывающей скобки и перед закрывающей, таким образом скобки отдаляются от текста, который они окружают и по сути не связываются с ним визуально.

Согласен, визуально плохо, надо исправить. Я это сделаю, просто пока не начинал, думаю справлюсь, просто ранее были другие задачи в парсере, сами сноски собрать в параграфы, смерджить, а скобки ушли на задний план, а сейчас можно переделать сноски, теперь уже понятно по эти скобкам, что мы хотим. В ближайшее время исправлю. Возможно это исправление можно объединить со склейкой коротких параграфов в один. подумаю, быстрее сноски сделать.

**Comment by iprst** [comment]
> Возможно это исправление можно объединить со склейкой коротких параграфов в один. подумаю, но, конечно, быстрее будет, если только сноски сделать. а короткие параграфы отложить

Однозначно лучше решить только сноски, поскольку они наиболее «острая проблема». Параграфы уже можно будет разобрать по накатанной схеме, даже возможно взяв что-то из полученного кода, если это будет актуальным.

**Comment by iprst** [comment]
По сноскам вопрос — а чём именно там самая ключевая загвоздка? На каком этапе процесса возникают причины текущего отображения?

**Comment by audetv** [comment]
> По сноскам вопрос — а чём именно там самая ключевая загвоздка? На каком этапе процесса возникают причины текущего отображения?

Я просто пока не начинал ещё ими заниматься. Никаких пока загвоздок нет, но только что надо сесть и сосредоточиться. Сделаю. Хотя есть загвоздка, просто css и шаблоны можно делать, отвлекаясь от рабочих дел, а в парсере не особо так получается. В ближайшие сутки сделаю)

**Comment by iprst** [comment]
А, нет, это не горит. Я о другом про загвоздку — может быть нам продумать более внятное их оформление, но без необходимости «переделывания всего». Потому и вопрос, где там собака порылась, чтобы понять что имеет смысл предлагать как варианты оформления. 

Например. 
- может это будет блок без параграфов <p> но с переносами строк <br>?
- может быть оформить другим шрифтом?
- может быть убрать скобки и сделать врезку типа как оформляют цитаты в таких блоках?
и тому подобное.

**Comment by iprst** [comment]
По поводу ссылок с номером параграфа. 
Возможно надо сделать всплывающую подсказку при наведении курсора, что-то вроде «посмотреть в контексте» или «посмотреть в контексте книги», какие-то такие варианты. Сейчас номера с оформлением ссылки работают почти как пасхалка — доступны только сообразительным.

**Comment by audetv** [comment]
> Сейчас номера с оформлением ссылки работают почти как пасхалка — доступны только сообразительным.

Ага, увидел в комментариях вопрос, да надо сделать по другому, согласен. Подумаю, но уже завтра, потестируем варианты.
Сегодня хочу отвлечься и почитать, что там в теме, я уже страниц 5 отстаю, столько событий, а я тут сижу примусы починяю)))

**Comment by iprst** [comment]
> Сегодня хочу отвлечься и почитать, что там в теме, я уже страниц 5 отстаю, столько событий, а я тут сижу примусы починяю)))

Да, конечно. Там была необходимость «протирать штаны».

**Comment by audetv** [comment]
Сделал тестовое обновление https://github.com/terratensor/kob-library-app/pull/34, можно оставить, можно вернуть как было. В режим 3 + 1 + 3.

Для книг ВП СССР можно применить такой режим чтения:
Сделал вверху параграфа ссылку с именем «Читать книгу»
При нажатии осуществляется переход в книгу к нужному параграфу.
Отображаются все параграфы. Книгу можно прочитать всю.
Что думаете?

**Comment by iprst** [comment]
Я думаю зря. Книги распространяются на сайтах КОБ в авторском виде. У нас эти файлы изменены для организации поиска, они не предназначались для публикации, только для поиска. Подразумевалось, что найдя нужную информацию пользователь возьмёт оригинальный авторский источник и прочитает.

**Comment by iprst** [comment]
С точки зрения пользовательского опыта — работает круто. Но за распространение книг будет прилёт.

**Comment by iprst** [comment]
Если распространять книги, то потребуется их форматирование под веб-типографику. Возможно они где-то опубликованы в текстовом виде, типа Wikireading? Я в открытом текстовом виде вообще не встречал, даже голый пдф не попадался. Всё лежит в архивах. Почему?

**Comment by audetv** [comment]
Наверное да, хорошо. 
Но это лишь про ВП СССР книги, не про другие библиотеки, с ними понятно, а в толстых книги в начале есть параграф, разрешающий «исходя из свойственного ему понимания общественной пользы копировать и тиражировать, в том числе с коммерческими целями, настоящие материалы в полном объеме или фрагментарно всеми доступными ему средствами», руководствовался этим. Но даже не  не в этом суть, я подумал, посмотрел, хорошо, но передумал) Есть в этом уход от направления в сторону, будет получается читалка (со своей спецификой), а нужен поисковый агрегатор)

**Comment by iprst** [comment]
> Но даже не не в этом суть, я подумал, посмотрел, хорошо, но передумал)

Почему? Не в смысле критики. Я чисто по безопасности маякнул — ситуация неизвестна (мне).

**Comment by audetv** [comment]
Т.е. ресурсов (время и сил) не так много, что бы покрыть все направления, нужно держаться главного вектора и не распыляться. Передумал, я так как начал делать, сделал и уже пошло по пфу давить на психику, в другую сторону нужны вебсокеты, сессии и прочее.  Пока не хочу) убрал. вернуть это не сложно, посмотрим, работы на полчаса, так как под это было сделано, есть база делай запрос и все)

**Comment by iprst** [comment]
Что думаете по геоматрице? Нереально?

**Comment by audetv** [comment]
Ахаха, понял, просто ещё мне хотелось почитать некоторые книги после разного поиска, вот и решил себе облегчить задачу). Ладно, вернул все как было с контекстом. но надо только вместо номеров придумать другую ссылку будет.

> Что думаете по геоматрице? Нереально?

Так нет, реально, но что бы это было реальнее, надо к ней переходить, а там как раз направление - цитаты из книг, и с эти все понятно и правильно, надо так и идти в этом векторе по книгам. Считаю, что геоматрица нужнее, ВП СССР книги уже есть в сети., а геоматрицы еще пока нет. так что будет двигаться.

**Comment by audetv** [comment]
Плюс задачи по геохешу впереди, по ней много чего интересного впереди, технических задач, поиск это лишь часть, геоматрицы. И она стала ближе мне, она уже сломала пару моих калейдоскопов, такое нельзя отпускать ))) Так что реально.

**Comment by iprst** [comment]
> Ахаха, понял, просто ещё мне хотелось почитать некоторые книги после разного поиска, вот и решил себе облегчить задачу).

Расчёт бессознательного понятен )))
Вероятно оно в правильную сторону ведёт, я обычно так предполагаю, типа это интуиция пробилась. Ну и хорошо — метод у вас уже отработан, далее его будет внедрить очень легко на любом подходящем материале. Если что, когда книги редактировал, то по таблице РКН с сайта вики-коб почти все опубликованные книги «зелёные» то есть проверенные, и только две или три «белые» то есть не проверялись (но скорее всего не будут и не содержат, ибо просто редакции других). 



> Так нет, реально, но что бы это было реальнее, надо к ней переходить, а там как раз направление - цитаты из книг, и с эти все понятно и правильно, надо так и идти в этом векторе по книгам.

Там по книгам только основная выдача «данных». Но в самой системе есть не отработанные механизмы, которые пока не понятно как будут работать. Основной проблемой выглядит комбинаторика, которая увеличивает сложность системы в _n!_ раз, а это плохой, нет, это наверное самый плохой вариант сложности.

**Comment by iprst** [comment]
> Плюс задачи по геохешу впереди

Со стороны логики программирования это наверняка наиболее близкая вам задача, поскольку это такая, ясная математика с понятными и логичными идеями. Чего не скажешь о координатах на геоиде ))

**Comment by audetv** [comment]
появилась мысль

По поводу ссылки на номере параграфа, при нажатии на ссылку, пока условно контекст (пока это номер параграфа, но от номеров надо 100% избавляться) 
осуществлять переход на страницу, сделанную в таком же режиме, как я делал в https://github.com/terratensor/kob-library-app/issues/20#issuecomment-1614984532
На странице будут 7 параграфов, и только в конце будет наименование книги, а не после каждого параграфа, как-бы страница контекста. Но чтобы номеров не было видно в поисковой строке. (возможно поисковая строка будет пустая, а это как бы страница контекста, preview страницы, большая циатата, надо подумать как ее назвать, чтобы было легально и правильно.

И так у нас не будет публичного размещения номеров параграфов и позиций параграфов относительно друг друга.

Надо подумать, но эти страницы можно сделать не доступными для индекса, например они каждый раз будут с новым кодом , и будут «протухать» пропадать/недоступны через какое-то время, чтобы нельзя было давать на эти страницы ссылку. И таким образом мы не распространяем, а лишь ознакамливаем читателя с книгой и направляем на путь)

**Comment by iprst** [comment]
> По поводу ссылки на номере параграфа, при нажатии на ссылку, пока условно контекст (пока это номер параграфа, но от номеров надо 100% избавляться)
> осуществлять переход на страницу, сделанную в таком же режиме, как я делал в [#20 (comment)](https://github.com/terratensor/kob-library-app/issues/20#issuecomment-1614984532)
> На странице будут 7 параграфов, и только в конце будет наименование книги, а не после каждого параграфа, как-бы страница контекста. Но чтобы номеров не было видно в поисковой строке. (возможно поисковая строка будет пустая, а это как бы страница контекста, preview страницы, большая циатата, надо подумать как ее назвать, чтобы было легально и правильно.

Мысль верная.
1. Если номера не нужны — убиваем номера, пишем «контекст», логика улучшается.
2. Правильная страница открывается в новой вкладке, там несколько параграфов подряд, как в книге. 
3. Внизу подвал с названием книги. Или вверху чердак, без разницы.
4. Чтобы не было номеров в строке — строку на странице делаем вообще невидимой, прячем её. 
5. Сделать раздел «для печати» и в роботс.тхт что-то вроде `disallow: /print_pages` 

> lib.svodd.ru/blockquote/ede125dc-19e3-11ee-be56-0242ac120002

Недоступно.

**Comment by audetv** [comment]
> 
> Мысль верная.
> 
> 1. Если номера не нужны — убиваем номера, пишем «контекст», логика улучшается.
> 2. Правильная страница открывается в новой вкладке, там несколько параграфов подряд, как в книге.
> 3. Внизу подвал с названием книги. Или вверху чердак, без разницы.
> 4. Чтобы не было номеров в строке — строку на странице делаем вообще невидимой, прячем её.
> 5. Сделать раздел «для печати» и в роботс.тхт что-то вроде `disallow: /print_page`
> 

Да, как то так. и да поисковую строку можно вообще убрать, в принципе да.

> > lib.svodd.ru/blockquote/ede125dc-19e3-11ee-be56-0242ac120002
> 
> Недоступно.

lib.svodd.ru/blockquote/ede125dc-19e3-11ee-be56-0242ac120002 - да недоступно, это был просто пример (такой страницы нет), как может выглядеть адрес такой страницы, uuid после bloслquote может быть всегда разным, для разных людей, или каждый раз разным при новом нажатии, над этим механизмом надо подумать. пока лишь направление, но как то так делают ссылки для скачивания, доступны некоторое время и потом пропадают.

**Comment by iprst** [comment]
> lib.svodd.ru/blockquote/ede125dc-19e3-11ee-be56-0242ac120002 - да недоступно, это был просто пример

А, подумал ссылка. 

> uuid после bloслquote может быть всегда разным, для разных людей, или каждый раз разным при новом нажатии, над этим механизмом надо подумать. пока лишь направление, но как то так делают ссылки для скачивания, доступны некоторое время и потом пропадают.

Значит в таком направлении и думаем.

**Comment by iprst** [comment]
Пока кнопка «контекст» в военно-исторической библиотеке работает не очень хорошо — страница /print_page пытается загрузиться в новом окне, но она очень долго пустая, прогружается около минуты. 

В случае, если страница контекста уже открыта, то при нажатии ссылки на контекст другого параграфа, его контекст открывается в той же вкладке что и предыдущий, и на это уходит в два раза меньше времени. Далее другие контексты загружаются в этой же вкладке примерно за полминуты. Если вкладку закрыть и снова нажать кнопку контекста, странице снова потребуется около минуты для загрузки. 

Сама страница выглядит адекватно, кажется это то, что доктор прописал.

**Comment by audetv** [comment]
Да, эта проблема спрозодительностью. Выборка происходит по uuid из 22 мил записей... Пока не очень хорошо. Думаю, что делать. проще всего увеличить мощность сервера, оперативную память.

**Comment by audetv** [comment]
Можно переписать запросы, можно переделать на выдачу страницы изтаостгресс, но надо понаблюдать, пробовать, в общем работа в переди)

**Comment by iprst** [comment]
> Да, эта проблема спрозодительностью. Выборка происходит по uuid из 22 мил записей...

Что если заранее сформировать таблицу «контекст» для каждой записи?

**Comment by iprst** [comment]
> Думаю, что делать. проще всего увеличить мощность сервера, оперативную память.

Представляется, что такой сценарий в будущем быстро превратится в метод «на каждую проблему выделять ресурсы сервера», кажется нужно исходить из оптимизации эффективности сценариев и наоборот понижать трату ресурсов.

**Comment by audetv** [comment]
Меньше чем сейчас, 2 гб памяти на все сервисы, врятли будет найден приемлемый сценарий. С большими данными. В самом начале комментарии работали на 512 мб, но падали если делать несколько запросов одновременно. 2 гб до этого было достаточно, а для таких больших данных уже нет но там помимо памяти ещё нужны диски) если не в памяти то на свопе. В общем так.

**Comment by iprst** [comment]
> Меньше чем сейчас, 2 гб памяти на все сервисы, врятли будет найден приемлемый сценарий. С большими данными. В самом начале комментарии работали на 512 мб, но падали если делать несколько запросов одновременно. 2 гб до этого было достаточно, а для таких больших данных уже нет но там помимо памяти ещё нужны диски)

Понял. Логично. Потому и был вопрос про ресурсы ещё до запуска. Но теперь знаем на практике, что для такого объёма данных текущих ресурсов мало, значит уже потребуется оценивать актуальность такого онлайн сервиса. Если, скажем, пользователей всего 5 (условно), то проще им раздать «локальное приложение» и сказать где база лежит ))
